Source/Core/Core/PowerPC/MMU.cpp

```diff
@@ -221,15 +224,28 @@ static T ReadFromHardware(u32 em_address)
     // Handle RAM; the masking intentionally discards bits (essentially creating
     // mirrors of memory).
     T value;
-    std::memcpy(&value, &Memory::m_pRAM[em_address & Memory::GetRamMask()], sizeof(T));
+    em_address &= Memory::GetRamMask();
+
+    if (!ppcState.m_enable_dcache || wi)
+      std::memcpy(&value, &Memory::m_pRAM[em_address], sizeof(T));
+    else
+      ppcState.dCache.Read(em_address, (u8*)&value, sizeof(T),
+                           HID0.DLOCK || flag != XCheckTLBFlag::Read);
+
     return bswap(value);
   }
 
   if (Memory::m_pEXRAM && (em_address >> 28) == 0x1 &&
       (em_address & 0x0FFFFFFF) < Memory::GetExRamSizeReal())
   {
     T value;
-    std::memcpy(&value, &Memory::m_pEXRAM[em_address & 0x0FFFFFFF], sizeof(T));
+    em_address &= 0x0FFFFFFF;
+
+    if (!ppcState.m_enable_dcache || wi)
+      std::memcpy(&value, &Memory::m_pEXRAM[em_address], sizeof(T));
+    else
+      ppcState.dCache.Read(em_address + 0x10000000, (u8*)&value, sizeof(T),
```

Comment:
Generally we add braces to both conditional branches in an if-else if the body of one of them trails more than one line. Ditto for the conditional above as well.

Also the `u8*` cast isn't necessary.